packages/cli/test/integration/controllers/invitation/invitation.controller.integration.test.ts

```diff
@@ -0,0 +1,422 @@
+import { mocked } from 'jest-mock';
+import Container from 'typedi';
+import { Not } from 'typeorm';
+
+import { InvitationController } from '@/controllers/invitation.controller';
+import { InternalHooks } from '@/InternalHooks';
+import { ExternalHooks } from '@/ExternalHooks';
+import { UserManagementMailer } from '@/UserManagement/email';
+import { UserRepository } from '@/databases/repositories/user.repository';
+import { PasswordUtility } from '@/services/password.utility';
+import { UserService } from '@/services/user.service';
+
+import {
+	randomEmail,
+	randomInvalidPassword,
+	randomName,
+	randomValidPassword,
+} from '../../shared/random';
+import { createMember, createOwner, createUserShell } from '../../shared/db/users';
+import { mockInstance } from '../../../shared/mocking';
+import * as utils from '../../shared/utils';
+
+import {
+	assertReturnedUserProps,
+	assertStoredUserProps,
+	assertUserInviteResult,
+} from './assertions';
+
+import type { User } from '@/databases/entities/User';
+import type { UserInvitationResult } from '../../shared/utils/users';
+
+describe('InvitationController', () => {
+	const testServer = utils.setupTestServer({ endpointGroups: ['invitations'] });
+
+	let instanceOwner: User;
+	let userRepository: UserRepository;
+	let externalHooks: jest.Mocked<ExternalHooks>;
+	let internalHooks: jest.Mocked<InternalHooks>;
+	let mailer: jest.Mocked<UserManagementMailer>;
+
+	beforeAll(async () => {
+		mailer = mockInstance(UserManagementMailer);
+		externalHooks = mockInstance(ExternalHooks);
+		internalHooks = mockInstance(InternalHooks);
+
+		userRepository = Container.get(UserRepository);
+
+		instanceOwner = await createOwner();
+
+		// selectively patch mocked dependencies into real services
+
+		Container.get(UserService).setMailer(mailer);
+		Container.get(InvitationController).setExternalHooks(externalHooks);
```

Comment:
I'd prefer not start using this format. All dependencies should be injected, and never modified after that.
To fix this test, we need to move all calls to `mockInstance` to be before `setupTestServer`, and that should do the trick.

Reply:
Thank you, I had a feeling this was not the way but could not find the answer in a short time.